src/array.rs

```diff
@@ -86,57 +87,169 @@ impl<'hw> Array<'hw> {
+    ///
+    /// # Arguments
+    ///
+    /// * `values` - Sequence of values to be set. The length must be the same as the size of
+    ///   the underlying buffer.
     ///
     /// # Returns
     ///
-    /// Scalar value of the array.
+    /// * `Ok(())` - Values are set correctly.
+    /// * `Err(Error)` - Length of the specified values is different with the size of the
+    ///   underlying buffer.
+    pub fn set_values_f32(&mut self, values: &[f32]) -> Result<()> {
+        if values.len() != self.shape.num_elements() {
+            return Err(Error::InvalidLength(format!(
+                "Values has a different length. Required {}, but got {}.",
+                self.shape.num_elements(),
+                values.len(),
+            )));
+        }
+
+        unsafe {
+            self.buffer.hardware().borrow_mut().copy_host_to_hardware(
+                values.as_ptr() as *const u8,
+                self.buffer.as_mut_handle(),
+                self.shape.num_elements() * mem::size_of::<f32>(),
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Obtains all values in the underlying buffer.
     ///
     /// # Returns
     ///
-    /// * `Ok(f32)` - Scalar value obtained from the array.
-    /// * `Err(Error)` - Array is not a scalar.
-    pub(crate) fn to_scalar(&self) -> Result<f32> {
-        self.shape.check_is_scalar()?;
-        let mut value = 0.;
+    /// `Vec` of all values. The order of values is row-major order (C order).
+    pub fn get_values_f32(&self) -> Vec<f32> {
+        let num_elements = self.shape.num_elements();
+        let mut values = Vec::<f32>::with_capacity(num_elements);
         unsafe {
             self.buffer.hardware().borrow_mut().copy_hardware_to_host(
                 self.buffer.as_handle(),
-                (&mut value as *mut f32) as *mut u8,
-                size_of::<f32>(),
+                values.as_mut_ptr() as *mut u8,
+                num_elements * mem::size_of::<f32>(),
             );
+            values.set_len(num_elements);
         }
-        Ok(value)
+        values
+    }
+
+    /// Creates a new `Array` with 0-dimensional shape on the specific hardware.
+    ///
+    /// # Arguments
+    ///
+    /// * `hardware`: `Hardware` object to host the value.
+    /// * `value`: Value of the resulting array.
+    ///
+    /// # Returns
+    ///
+    /// A new `Array` object representing a scalar value.
+    pub fn scalar_f32(hardware: &'hw RefCell<dyn Hardware>, value: f32) -> Self {
+        let mut array = unsafe { Self::raw(hardware, make_shape![]) };
+        array.set_scalar_f32(value).unwrap();
+        array
+    }
+
+    /// Creates a new `Array` with arbitrary shape on the specific hardware.
+    ///
+    /// # Arguments
+    ///
+    /// * `hardware`: `Hardware` object to host the value.
+    /// * `shape`: `Shape` of the new `Array`.
+    /// * `value`: Values copied to the underlying buffer. The number of values must be the same as
```

Comment:
nit: I was a bit confused about when `values` are copied to the buffer.

```suggestion
    /// * `values`: Values to be copied to the underlying buffer. The number of values must be the same as
```